Add image block aspect ratio control

ajlende opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Adds a new dimensions controller that includes aspect-ratio, object-fit (size), width, and height.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Fixes #51138

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- Adds a new component in the @wordpress/block-editor package and exports it in the private APIs.
- Imports and uses the component in the image block.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Storybook

1. Open storybook (`npm run storybook:dev`).
2. New components added under "Block Editor (Private APIs)".

### Image Block

1. Create an image block in the post editor.
2. Use the tool in the sidebar.
3. Resize handles, centering, border, image cropper all have interactions with the width/height attribute. Test them out with the changes.
4. Try the block deprecations by opening a post with an old image block.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

Same as above, with the keyboard.

## Screenshots or screencast <!-- if applicable -->

![Post editor aspect ratio tool](https://github.com/WordPress/gutenberg/assets/5129775/599e4ad7-3811-435e-a65e-ca84cf760a8d)

Commit:
Simplify ImageSizeControl by using Auto as a placeholder

Commit:
Rename imageWidth and imageHeight props to naturalWidth and naturalHeight

Commit:
Convert NumberControl onChange values to Numbers

Commit:
Simplify LatestPostsEdit to use updated ImageSizeControl

Commit:
Add JSDoc types for debugging

Commit:
Remove unnecessary noop

Commit:
Fix possible undefined values in NumberControl onChange

Commit:
Fix onChangeImage param type which may be undefined

Commit:
Rename OnChange callback prop

Commit:
Inline JSDoc props instead of new object

Commit:
Simplify handing undefined and NaN in onChange

Commit:
Revert prop name change since this isn't a private API

Commit:
Add a privateApis export for experimental ImageSizeControl

Commit:
Use the privateApis version of ImageSizeControl

Commit:
Add deprecation notice to the original component

Commit:
Revert image-size-control and create image-dimensions-control instead

Commit:
Re-add deprecation notice to image-size-control

Commit:
Try making a whole new component

Commit:
Revert changes to image, latest-posts, and media-text blocks

Commit:
Organize and update the dimensions tool panel item

Commit:
Reword size help text

Commit:
Reorganize into reusable components

Commit:
Add stories for other individual tools

Commit:
Update stories path

Commit:
Remove SelectControl __next prop

Commit:
Pass through isShownByDefault to ResolutionTool

Commit:
Remove unused scss

Commit:
Deprecate experimental ImageSizeControl

Commit:
Simplify ScaleTool onChange

Commit:
Add better defaults for value and onChange

Commit:
Fix circular dependency

Commit:
Update comment about auto and custom aspect ratios

Commit:
Add JSDoc types for ScaleTool

Commit:
Add JSDoc types for WidthHeightTool

Commit:
Add default value and onChange for WidthHeightTool

Commit:
Remove unused import

Commit:
Add aspectRatio to image block attributes

Commit:
Add scale to image block attributes

ajlende committed:
Update JSDoc comment

ajlende committed:
Add dimensions tool to image block

ajlende committed:
Rename naturalAspectRatio for clarity

ajlende committed:
Fix aspect-ratio-tool lint

ajlende committed:
Fix scale-tool lint

ajlende committed:
Fix width-height-tool lint

ajlende committed:
Fix dimensions-tool lint

ajlende committed:
Fix resolution-tool lint

ajlende committed:
Add @emption/styled to block-editor

ajlende committed:
Fix image block lint

ajlende committed:
Update components changelog

ajlende committed:
Fix AspectRatioTool reference

ajlende committed:
Support 'auto' in width-height-tool

ajlende committed:
Make null/undefined values mean 'auto' instead of defaultValue in aspectRatioTool

ajlende committed:
Add deprecation for image block

ajlende committed:
Fix ResizableBox interactions

ajlende committed:
Add comments for default values

ajlende committed:
Fix ResizableBox with auto w/h

ajlende committed:
Clear aspect-ratio on resize

ajlende committed:
Add TODO comment for ResolutionTool defaultValue

ajlende commented:
There are still a few TODOs in the comments and tests need to be updated, but I wouldn't mind some early reviews, so I've opened this up as ready for review.

jasmussen commented:
This is really great, thanks for the PR! Let's see if we can't get this in 6.3 it's so close. Here's a GIF showing the aspect ratio dropdown, how it affects scaling after the fact, the fill control, and even a custom aspect ratio when both width and height are entered:

![aspect ratio](https://github.com/WordPress/gutenberg/assets/1204802/d9216015-2051-4d67-b71c-fd4e802243cb)

There is one thing that I would like us to fix, but which shouldn't be blocking for the PR. Here's a testimonial pattern with a circular avatar that I insert. Note two things:

* When I insert this pattern fresh, it shows up as having a _custom aspect ratio_ even though I would expect it to have the _original_ aspect ratio.
* Also note that with this pattern, both in trunk and here, if I save and reload the particular pattern shows up as having invalid block content.

So it seems like there's an issue with that separate pattern. But nevertheless even when recovered, there's an issue we should fix, which is that when a custom aspect ratio is applied, _dimensions and aspect ratio should be retained across image replacements_. In the following GIF I drop a big landscape image onto the small circular avatar:

![replace flow](https://github.com/WordPress/gutenberg/assets/1204802/e8a4ca89-7694-41c1-9554-0de6f7744358)

Notice how the small circular avatar becomes a bit pill shape instead. Ideally, and this can be separate, the dimensions should've been retained when the custom aspect ratio is selected.

---

A few small style tweak. When there's no segmented control for scale mode, there's too much space below the aspect ratio dropdown:

<img width="286" alt="Screenshot 2023-06-20 at 09 50 41" src="https://github.com/WordPress/gutenberg/assets/1204802/60367191-8d39-4735-be3e-31e5a00959f4">

I personally think that the "fill" mod of this segmented control is completely useless and isn't something we should surface to users:

<img width="123" alt="Screenshot 2023-06-20 at 09 51 03" src="https://github.com/WordPress/gutenberg/assets/1204802/f44356e5-ce43-41e1-80e4-0950b2c38f29">

I realize it's already shipping and for featured image too. But could we / should we remove that option?

getdave commented on packages/block-editor/src/components/dimensions-tool/index.js:
```diff
@@ -0,0 +1,182 @@
+	// dropdown to indicate that setting both the width and height is the same
+	// as a custom aspect ratio.
+	const aspectRatioValue =
+		value.width !== null &&
+		value.width !== undefined &&
+		value.height !== null &&
+		value.height !== undefined
```
As these tests happen multiple times it might be worth extracting a well named `const`.

getdave commented on packages/block-editor/src/components/dimensions-tool/index.js:
```diff
@@ -0,0 +1,182 @@
+					if (
+						nextAspectRatio !== 'custom' &&
+						nextAspectRatio !== 'auto' &&
+						nextAspectRatio !== null &&
+						nextAspectRatio !== undefined
```
Nit: the pattern of checking `null || undefined` could be extracted to an `isNil` util. Might help cut down the verbosity.

ajlende replied:
JavaScript has an operator that does exactly that supported all the way back to the inception of the language, if only we were allowed to use it...

That being said, I do want to clean this up, but probably by settings things with a mixture of ternary and nullish coalescing operators to flatten the values into one before checking.

scruffian replied:
>  if only we were allowed to use it...

why aren't we?

ajlende replied:
> why aren't we?

https://github.com/WordPress/gutenberg/blob/82d99b94263ddcaa9cdd6fb615ed331f00fdbbe6/packages/eslint-plugin/configs/es5.js#L19

I guess you could technically do this.

```js
/* eslint-disable eqeqeq */
...
nextAspectRatio != null
...
/* eslint-enable eqeqeq */
```

getdave commented on packages/block-library/src/image/utils.js:
```diff
@@ -3,6 +3,15 @@
  */
 import { NEW_TAB_REL } from './constants';
 
+export function evalAspectRatio( value ) {
```
```suggestion
export function calculateAspectRatio( value ) {
```

Nit: the word "eval" in JavaScript was a trigger for me 😄 I'm not sure we're evaluating here anyway - more "calculating".

ajlende replied:
I guess I chose `eval` because it's doing the same thing as the JS `eval()` taking some source code string (in our case a CSS aspect-ratio value) and returning the numerical aspect ratio. `eval('5/8')`, for example, does the same thing.

I'd expect something with `calculate` in the name to take a numerical `width` and `height` argument instead of a single string.

getdave replied:
Ah. You're right. It's a string. I missed that. Moving too quickly 👍

getdave commented on packages/block-editor/src/components/dimensions-tool/scale-tool.js:
```diff
@@ -0,0 +1,113 @@
+		value: 'contain',
+		label: _x( 'Contain', 'Scale option for dimensions control' ),
+		help: __(
+			'Content is scaled to fill the space without clipping nor distorting.'
```
```suggestion
			'Content is scaled to fill the space without clipping or distorting.'
```

ajlende replied:
This was copied over from the post featured image block. Eventually I'd like to replace that component with this one, so I'm not going to bother changing it over there for now.

getdave commented on packages/block-editor/src/components/dimensions-tool/scale-tool.js:
```diff
@@ -0,0 +1,113 @@
+	const displayValue = value ?? 'fill';
+
+	// TODO: Since we're passing in options, would it make sense to hardcode the help text as a lookup for all CSS values inste?
+	const scaleHelp = useMemo( () => {
+		return options.reduce( ( acc, option ) => {
+			acc[ option.value ] = option.help;
+			return acc;
+		}, {} );
+	}, [ options ] );
```
I guess it's a trade off on your TODO question.

The overhead of another prop and the consumer having to avoid typos in the lookup, versus the potential performance if the user passes an unstable `options` reference.

Given the size of the component subtree it should be fine to leave it as it is though.

getdave commented on packages/block-editor/src/components/dimensions-tool/scale-tool.js:
```diff
@@ -0,0 +1,113 @@
+			onDeselect={ () => onChange( defaultValue ) }
+			panelId={ panelId }
+		>
+			{ showControl && (
```
If this only displays when `showControl` is truthy then why not return early and avoid rendering the `ToolsPanelItem`? Is it a reconciliation optimisation? Ie

ajlende replied:
The `ToolsPanelItem` registers the reset button for the scale (three dots menu). It also adds a wrapping `div` that causes the extra space that @jasmussen mentioned.

I made the tradeoff because it seemed weird to me that the reset button (which in other contexts is used to show the tool if hidden by default) didn't exist unless you selected an aspect ratio.

draganescu commented on packages/block-editor/src/components/dimensions-tool/aspect-ratio-tool.js:
```diff
@@ -0,0 +1,124 @@
+ */
+
+/**
+ * These should use the same values as AspectRatioDropdown in @wordpress/block-editor.
```
It would be good to add to the comment why the duplication is needed and can't be DRY.

ajlende replied:
It isn't that they can't be DRY, but this PR is already big enough and I'd have to update the post featured image block too. They also technically aren't the same right now either, so maybe a TODO comment that gets committed is more appropriate.

ajlende committed:
Move the scale hide/show into dimensions controls

ajlende committed:
Add first test

ajlende committed:
Fix scale being set after it was deleted

ajlende committed:
WIP writing tests

ajlende committed:
Update test

richtabor commented:
> which is that when a custom aspect ratio is applied, dimensions and aspect ratio should be retained across image replacements. 

I also think it would be good to maintain aspect ratio (along with height and width) if an image is replaced. If there's no aspect ratio, then sure a complete replacement is good to go, but if there is an aspect ratio, it should stick.

Commit:
UI tweaks

Commit:
Move alt text as ToolsPanelItem

Commit:
Tweak default scale option help text

Commit:
Only use contain and cover for image scale options

ajlende committed:
Update test

ajlende committed:
Test the remaining callback values

ajlende committed:
Add comment about toStrictEqual

ajlende committed:
Add test for setting custom aspect ratio and then resetting

ajlende committed:
Move custom scaleOptions to the image block

ajlende committed:
Remember last aspect ratio so it can be restored when with/height are unset then set

ajlende committed:
Remove unused import

ajlende committed:
Format code

ajlende committed:
Remove image w/h reset when a new image is added

ajlende commented:
Thanks for the review @jasmussen. I fixed a few things today, but I'll be continuing on with the rest tomorrow.

> when a custom aspect ratio is applied, dimensions and aspect ratio should be retained across image replacements.

Fixed in 0075ec7

> A few small style tweak. When there's no segmented control for scale mode, there's too much space below the aspect ratio dropdown

Fixed in 0e14ad6

> I personally think that the "fill" mod of this segmented control is completely useless and isn't something we should surface to users... I realize it's already shipping and for featured image too. But could we / should we remove that option?

Fixed in a44ae44

This seems like something that would be good to change in the Post Featured Image block as part of that follow-up too.

jasmussen reviewed (approved):
Fantastic, let's hustle to get this in. Just for reference, the fact that patterns can retain their original dimensions is solid.

richtabor commented:
I think this works well!

ajlende commented:
I wanted to make sure that I had all the transitions covered, so I made a few diagrams to figure out everything that was needed. The last one is the important one, and I'll be writing tests for all the blue edges.

![Dimensions_Tool_FSM_1](https://github.com/WordPress/gutenberg/assets/5129775/8f2f31c2-e070-47d0-b893-91d71cd4c814)
![Dimensions_Tool_FSM_2](https://github.com/WordPress/gutenberg/assets/5129775/60ed55c7-d38f-4dc3-ac26-cb40d8f3c275)
![Dimensions_Tool_FSM_3](https://github.com/WordPress/gutenberg/assets/5129775/2461b56d-676c-4410-aeea-d7828e4d537d)
![Dimensions_Tool_FSM_4](https://github.com/WordPress/gutenberg/assets/5129775/88b4e418-9b91-49e5-9852-467ec9d980a7)

[Source LaTeX](https://www.overleaf.com/read/gdngqyhhktyj) since I often get asked how I make these diagrams.

A few more notes about these graphs:

The asymmetry is caused by 2 things: scale is always automatically toggled because it has no `null` state, and `h` is preferentially cleared when you would enter an invalid state while entering an aspect ratio.

`1011` is the only state that is not an intermediary state because it requires setting two values at once to reach.

In graphs like the first one where one bit is toggled at a time, the number of edges grows at a rate of `f(n) = n*2^n` ([OEIS A036289](https://oeis.org/A036289)). The number of edges in each layer is twice the denominators in Leibniz's Harmonic Triangle ([OEIS A003506](https://oeis.org/A003506)). You probably won't want to do this exercise with n>5.

ajlende committed:
Use UnitControl's default units instead of spacing.units

ajlende committed:
Provide the complete set of object-fit options by default

ajlende committed:
Update TODO that will be committed

ajlende committed:
Clean up evalAspectRatio and add docs

ajlende committed:
Someone can file a bug report if offsetWidth/offsetHeight causes issues

ajlende committed:
I couldn't figure out why height depended on having a custom border, but things seem to work without that

ajlende committed:
Update docs for image block

ajlende committed:
Update comment about default value

ajlende committed:
Fix redundant wording

ajlende committed:
I think the img width and height attributes can be removed if they're specified in the style attribute

ajlende committed:
Update package-lock.json with @emotion/styled dependency

ajlende committed:
Update mock calls for test example

ajlende committed:
Simplify test values

ajlende committed:
Consolidate mock calls expect

ajlende committed:
Require defaultScale and defaultAspectRatio for DimensionsTool

ajlende committed:
Add DimensionsTool tests for all custom transitions

ajlende committed:
Remove comment about matching aspect ratio options

ajlende committed:
Remove redundant check in tests

ajlende committed:
Add comments to defaultAspectRatio and defaultScale

ajlende committed:
Organize tests by which field is being updated

ajlende committed:
Fix type conversion

ajlende committed:
Add state diagram for last two tests

ajlende committed:
Refactor and fix some tests

ajlende committed:
Fix and simplify WidthHeightTool onChange

ajlende committed:
Remove default scale option in image block.json

ajlende committed:
Simplify DimensionsTool onChange logic

ajlende committed:
Update block deprecations with width and height

ajlende committed:
Revert image block width and height attributes to numbers since we only support px units for now

ajlende committed:
Revert "Update block deprecations with width and height"

This reverts commit 941a81149ed4bc344ac2c0e183624069e33d75ad.

ajlende committed:
Prevent NaN width/height

ajlende committed:
Fix DimensionTool width/height units

ajlende committed:
Fix JSDoc Dimenstions width/height types

ajlende committed:
No default needed for ResolutionTool

richtabor reviewed (approved):
Very nice ✨

ajlende committed:
Fix drag handle aspect ratio reset

ajlende committed:
Simplify null checks

ajlende committed:
Stop using pxWidth and pxHeight

richtabor commented:
Works, and looks, great. 🚀

https://github.com/WordPress/gutenberg/assets/1813435/221ebb22-24c0-4c6a-9994-4dbacdc15f34

ajlende committed:
Remove e2e tests that reference the scale button that was removed

ajlende committed:
Fix image scaling for small images

ajlende committed:
Try fixing aspectRatio only images

jeryj committed:
Update test to respect the new aspect ratio behavior

ajlende commented:
The remaining failing e2e tests are unrelated to this change. We should be able to merge now.

alessandrocarrera commented:
This feature is amazing! When will it be available?

jasmussen commented:
🚀

ajlende commented:
> This feature is amazing! When will it be available?

@alessandrocarrera I think we merged it just in time for it to be in WP 6.3

stokesman commented on test/e2e/specs/editor/blocks/buttons.spec.js:
```diff
@@ -150,28 +150,6 @@ test.describe( 'Buttons', () => {
 		);
 	} );
 
-	test( 'can resize width', async ( { editor, page } ) => {
-		await editor.insertBlock( { name: 'core/buttons' } );
-		await page.keyboard.type( 'Content' );
-		await editor.openDocumentSettingsSidebar();
-		await page.click(
-			`role=region[name="Editor settings"i] >> role=tab[name="Settings"i]`
-		);
-		await page.click(
-			'role=group[name="Button width"i] >> role=button[name="25%"i]'
-		);
-
-		// Check the content.
-		const content = await editor.getEditedPostContent();
-		expect( content ).toBe(
-			`<!-- wp:buttons -->
-<div class="wp-block-buttons"><!-- wp:button {"width":25} -->
-<div class="wp-block-button has-custom-width wp-block-button__width-25"><a class="wp-block-button__link wp-element-button">Content</a></div>
-<!-- /wp:button --></div>
-<!-- /wp:buttons -->`
-		);
-	} );
-
```
> Remove e2e tests that reference the scale button that was removed

This test, being for the Buttons block, seems to have been removed in haste. Excuse me if I am missing something though.

ajlende replied:
My bad. I searched for `role=button[name="25%"i]` that was removed in the image block and mistakenly removed it here. Opened https://github.com/WordPress/gutenberg/pull/51865 to restore it.

tellthemachines commented:
I just cherry-picked this PR to the release/16.1 branch to get it included in the next release: d04f19fc36

porg commented:
@jasmussen wrote
> Notice how the small circular avatar becomes a bit pill shape instead. Notice how the small circular avatar becomes a bit pill shape instead. Ideally, and this can be separate, the dimensions should've been retained when the custom aspect ratio is selected.

**General Hint:** To preserve a circular shape no matter a 64x64px or 640x320 or a 1980x1080 image: Simply set the border radius to some huge value like 10000px and all your shapes will be circular.

**To preserve border-radiuses proportional to the image dimensions:** I have not tested this, but possibly percent values are the solution? If the top-left and top-right border-radius values are e.g. 10% then a 600px wide image has a rounding of 60px, a 1200px wide image a rounding of 120px. Or you could set the border-radiuses in relation to the viewport with `vw` and `vh` possibly.

jasmussen commented:
Just to clarify, the intent here was to preserve the _dimensions_ of the image even when dropping in a replacement, so the metrics of a pattern are retained even if you update the imagery inside. That's addressed here, and working really well 👌

porg commented:
Thanks for your clarification:
- My border-radius-preservation hints were well intended but seemingly off-topic. Maybe of help to you elsewhere 😉
- Now I understand your complaint on the "pill shape" was mainly a complaint about a square getting a landscape rectangle, and the rounded edges was just a side effect, which you mentioned, but which was not part of the complaint.

jasmussen commented:
👌

anomiex commented on packages/block-editor/package.json:
```diff
@@ -32,6 +32,7 @@
 	],
 	"dependencies": {
 		"@babel/runtime": "^7.16.0",
+		"@emotion/styled": "^11.6.0",
```
This results in an unexpected transitive peer dependency on `@emotion/react`. See #52474 for details.

ramonjd commented on packages/block-library/src/image/image.js:
```diff
@@ -504,25 +564,20 @@ export default function Image( {
 				className={ borderProps.className }
-				style={ borderProps.style }
+				style={ {
+					width:
```
Did we need a deprecation here?

It's causing some deprecation errors on existing blocks that do not have the styles set.

See https://github.com/WordPress/gutenberg/issues/52388#issuecomment-1631721808

ajlende replied:
@ramonjd This is part of `edit`, not `save`, so it shouldn't affect block deprecations. In a follow-up https://github.com/WordPress/gutenberg/pull/52286, @draganescu added styles to `save`, but he did add a deprecation for them.

ramonjd replied:
Doh! My bad for overseeing it. Thanks for the quick response 🙇🏻

Mamaduka commented on packages/block-editor/src/components/image-size-control/index.js:
```diff
@@ -30,6 +31,11 @@ export default function ImageSizeControl( {
 } ) {
+	deprecated( 'wp.blockEditor.__experimentalImageSizeControl', {
+		since: '6.3',
+		alternative:
+			'wp.blockEditor.privateApis.DimensionsTool and wp.blockEditor.privateApis.ResolutionTool',
```
I just noticed this message when testing the "Media & Text" block, which still uses the deprecated component.

This `alternative` suggestion isn't helpful for third-party consumers since the private APIs aren't available for them. Should we remove this from the deprecation message?

fabiankaegy replied:
Also just noting here since I just ran into this yesterday, it isn't fun for third party extenders when a component is deprecated whilst the replacement is still marked as private 😞

andrewserong commented:
I just noticed a bug with resetting the alt text field, that I think might have been introduced in this PR. I have a fix up over in #56809.

jasmussen commented:
@andrewserong only since I'm capturing you here, if you get a moment of time, can you look at whether #54032 is something that could be fixable? I'd love to pair up with you (or anyone who has time) to work on that one. It's no more urgent than everything else, so this is only an appeal in case it's workable.

andrewserong commented:
Thanks for the ping, Joen! That does sound like a very cool feature, I'll add a couple of comments over on #54032.
